pallets/parachain-staking/src/lib.rs

```diff
@@ -444,94 +621,128 @@ pub mod pallet {
 		pub fn go_online(&mut self) {
 			self.state = CollatorStatus::Active;
 		}
-		pub fn leave(&mut self, when: RoundIndex) {
+		pub fn leave<T: Config>(&mut self) -> Result<(RoundIndex, RoundIndex), DispatchError> {
+			ensure!(!self.is_leaving(), Error::<T>::CandidateAlreadyLeaving);
+			let now = <Round<T>>::get().current;
+			let when = now + T::LeaveCandidatesDelay::get();
 			self.state = CollatorStatus::Leaving(when);
+			Ok((now, when))
 		}
 	}
 
-	impl<A: Clone, B: Copy> From<Collator2<A, B>> for CollatorSnapshot<A, B> {
-		fn from(other: Collator2<A, B>) -> CollatorSnapshot<A, B> {
+	impl<A: Clone, B: Copy> From<CollatorCandidate<A, B>> for CollatorSnapshot<A, B> {
+		fn from(other: CollatorCandidate<A, B>) -> CollatorSnapshot<A, B> {
 			CollatorSnapshot {
 				bond: other.bond,
-				nominators: other.top_nominators,
+				delegations: other.top_delegations,
 				total: other.total_counted,
 			}
 		}
 	}
 
 	#[derive(Clone, PartialEq, Encode, Decode, RuntimeDebug, TypeInfo)]
-	pub enum NominatorStatus {
+	pub enum DelegatorStatus {
 		/// Active with no scheduled exit
 		Active,
-		/// Schedule exit to revoke all ongoing nominations
+		/// Schedule exit to revoke all ongoing delegations
 		Leaving(RoundIndex),
 	}
 
 	#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
-	/// Nominator state
-	pub struct Nominator2<AccountId, Balance> {
-		/// All current nominations
-		pub nominations: OrderedSet<Bond<AccountId, Balance>>,
-		/// Nominations scheduled to be revoked
-		pub revocations: OrderedSet<AccountId>,
-		/// Total balance locked for this nominator
+	/// Delegator state
+	pub struct Delegator<AccountId, Balance> {
+		/// Delegator account
+		pub id: AccountId,
+		/// All current delegations
+		pub delegations: OrderedSet<Bond<AccountId, Balance>>,
```

Comment:
This is maybe more of a note to myself to verify later, but does this `OrderedSet` need to be re-sorted upon migration?

Reply:
Nevermind, I don't think this question makes sense, but I still want to revisit it

Reply:
No need to resort. The ordering isn't actually used anywhere, the OrderedSet is used to enforce the set constraint that there can be no two delegations to the same candidate. A better data structure could be used, definitely, but not in scope for this PR.

It'd be great to remove OrderedSet altogether, but it would require migrating a few storage items.